[camera] Move camera streaming to platform interface

stuartmorgan opened the pull request:
The image streaming code was still embedded directly into the app-facing package, which prevents federation of the implementation and transition to per-platform method channels, as well as preventing other platforms from implementing it. It was originally not moved because there was discussion of changing to a completely different approach to image streaming, but there's no short-term plan to revisit that, so we should put the actual implementation we have into the platform interface to support federation. Eventually if we redesign streaming, we can do a breaking change.

To avoid a breaking change in `camera` for now, duplicate data types for the streaming images are defined in the platform interface layer, and they are translated to the existing classes in the app-facing layer. Since they are simple data classes, this can be eliminated in a future breaking change to the app-facing package; a TODO in the code captures that.

Part of https://github.com/flutter/flutter/issues/100555

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

bparrishMines commented on packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart:
```diff
@@ -149,6 +149,21 @@ abstract class CameraPlatform extends PlatformInterface {
+  ///
+  // TODO(bmparr): Add options to control streaming settings (e.g.,
+  // resolution and FPS).
+  Stream<CameraImageData> onStreamedFrameAvailable(int cameraId,
```
I'm confused by this name. Does the app-facing package notify the platform implementation package that a frame is available? Based on the `MethodChannel` implementation, it looks like this method creates the stream that the app-facing package listens to. Would a better name be something like `createCameraImageDataStream`?

stuartmorgan replied:
I don't like this name either; I used this name because it's the pattern used by [all of the existing Streams in the interface](https://github.com/flutter/plugins/blob/main/packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart#L70-L100)

Do you want to pick a new naming scheme for platform interface streams in this plugin as part of this PR? If so I could make alternate named versions of all of the existing ones, doing passthroughs to the old names (which would be soft-deprecated with a comment), so that we'd get consistency.

bparrishMines replied:
Ah, I see. I think I was confused by the call to [startImageStream](https://github.com/flutter/plugins/pull/5783/files#diff-8d80e3b1252c652ffc58ad176f59eb21263bf94dee032821ea7d61b917a6bca8R295). I understand now that the app-facing package is adding a listener to the stream which is created by the platform implementation. I think the current name is appropriate for how it is being used then. Streams and callbacks in Flutter tend to use the `on<event-name>` pattern like the one here.

stuartmorgan commented on packages/camera/camera/lib/src/camera_image.dart:
```diff
@@ -7,11 +7,23 @@ import 'dart:typed_data';
 import 'package:camera_platform_interface/camera_platform_interface.dart';
 import 'package:flutter/foundation.dart';
 
+// TODO(stuartmorgan): Remove all of these classes in a breaking change, and
```
Good point, I filed https://github.com/flutter/flutter/issues/104188

I'm not sure if we want to do a breaking change just for this or not. It's pretty minor... but on the other hand most clients wouldn't be affected, and it's a quick change for those who are, so it wouldn't be too disruptive to do.

stuartmorgan commented on packages/camera/camera/test/camera_image_test.dart:
```diff
@@ -5,11 +5,64 @@
+
+    final CameraImage image = CameraImage.fromPlatformInterface(originalImage);
+    // Simple values.
+    expect(image.width, originalImage.width);
```
Done.

stuartmorgan commented on packages/camera/camera_platform_interface/lib/src/types/camera_image_data.dart:
```diff
@@ -0,0 +1,126 @@
+  /// The distance between adjacent pixel samples in bytes, when available.
+  final int? bytesPerPixel;
+
+  /// Height of the pixel buffer, when available
```
Fixed.

stuartmorgan commented on packages/camera/camera_platform_interface/lib/src/types/camera_image_data.dart:
```diff
@@ -0,0 +1,126 @@
+  /// Height of the pixel buffer, when available
+  final int? height;
+
+  /// Width of the pixel buffer, when available
```
Fixed.

stuartmorgan commented on packages/camera/camera_platform_interface/test/method_channel/method_channel_camera_test.dart:
```diff
@@ -1038,6 +1038,52 @@ void main() {
       });
+
+      test('Should start streaming', () async {
+        // Arrange
```
I'm not a big fan of the commenting myself, although I can see the value in reminding people to always think this way in test structure (since it is a good structure, and I've seen plenty of tests that turn into soup when people don't follow it). I was just going for file-level consistency.

bparrishMines reviewed (approved):
LGTM

stuartmorgan commented:
I was going to make a new PR with just the platform interface parts per the usual process, but discovered that I named this branch as if it were going to be that PR, and then accidentally made it the combined PR, so I've reverted the `camera` changes, I'll land this as the platform interface change, and I've moved the combined PR to https://github.com/flutter/plugins/pull/5833.
